### PR TITLE
Simplify information in IncrementalIndex

### PR DESCRIPTION
**Pull request by navis**
For
```
private final Map<String, Integer> metricIndexes;
private final Map<String, String> metricTypes;
private final ImmutableList<String> metricNames;
```
and
```
private final LinkedHashMap<String, Integer> dimensionOrder;
private final DimensionHolder dimValues;
private final CopyOnWriteArrayList<String> dimensions;
```
each group can be merged into single map

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 1]
```diff
@@ -244,22 +243,57 @@ public int lookupId(String name)
...
+
   private final long minTimestamp;
   private final QueryGranularity gran;
   private final List<Function<InputRow, InputRow>> rowTransformers;
-  private final AggregatorFactory[] metrics;
-  private final Map<String, Integer> metricIndexes;
-  private final Map<String, String> metricTypes;
-  private final ImmutableList<String> metricNames;
-  private final LinkedHashMap<String, Integer> dimensionOrder;
+  private final AggregatorFactory[] factories;
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: why change this from metrics?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 2]
```diff
@@ -244,22 +243,57 @@ public int lookupId(String name)
     };
   }
 
+  public static class DimensionDesc
+  {
+    final int index;
+    final String name;
+    final DimDim values;
+    final ColumnCapabilitiesImpl capabilities;
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: can we create getters for these variables and make them private?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 3]
```diff
@@ -244,22 +243,57 @@ public int lookupId(String name)
...
+      this.capabilities = capabilities;
+    }
+  }
+
+  public static class MetricDesc
+  {
+    final int index;
+    final String name;
+    final String type;
+    final ColumnCapabilitiesImpl capabilities;
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: same comment here

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 4]
```diff
@@ -244,22 +243,57 @@ public int lookupId(String name)
...
-  private final Map<String, ColumnCapabilitiesImpl> columnCapabilities;
   private final boolean deserializeComplexMetrics;
 
-  protected final CopyOnWriteArrayList<String> dimensions;
+  private final Map<String, MetricDesc> metrics;
+  private final Map<String, DimensionDesc> dimensions;
+  private final Map<String, ColumnCapabilitiesImpl> columnCapabilities;
 
-  private volatile AtomicInteger numEntries = new AtomicInteger();
+  private transient final AtomicInteger numEntries = new AtomicInteger();
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: why the need for this change?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 5]
```diff
@@ -287,54 +321,34 @@ public IncrementalIndex(
...
+    for (String dimension : dimensionsSpec.getDimensions()) {
       ColumnCapabilitiesImpl capabilities = new ColumnCapabilitiesImpl();
       capabilities.setType(ValueType.STRING);
+      dimensions.put(dimension, new DimensionDesc(dimensions.size(), dimension, newDimDim(dimension), capabilities));
       columnCapabilities.put(dimension, capabilities);
-      dimensionOrder.put(dimension, dimensionOrder.size());
-      dimValues.add(dimension);
+    }
+
+    // This should really be more generic
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: i agree :)

**Comment by fjy** [comment]
I'm in favor of this PR but I have some outstanding comments

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 1 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: I used `metrics` for metrics. Ok, I'll revert that.

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 2 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: sure

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 3 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: ok

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 4 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: What will be changed is the value of `numEntries` not `numEntries` itself. Need not to be volatile at all, doesn't it?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 6]
```diff
@@ -244,22 +243,97 @@ public int lookupId(String name)
     };
   }
 
+  public static class DimensionDesc
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: The unofficial convention is to put static classes at the end of class definitions, if I recall from @himanshug 's formatting requests on other PRs

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 7]
```diff
@@ -419,32 +481,38 @@ public int add(InputRow row) throws IndexSizeExceededException
...
+        if (desc == null) {
+          desc = new DimensionDesc(dimensionDescs.size(), dimension, newDimDim(dimension), capabilities);
+          dimensionDescs.put(dimension, desc);
 
           if (overflow == null) {
             overflow = Lists.newArrayList();
           }
-          overflow.add(getDimVals(dimValues.add(dimension), dimensionValues));
-        } else if (index > dims.length || dims[index] != null)  {
+          overflow.add(getDimVals(desc.values, dimensionValues));
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: use getter

**Comment by drcrallen** [comment]
Very reasonable and easier to read changes. Can you use the getters for the class fields instead of the fields directly?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 6 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: ok

**Comment by nishantmonu51** [comment]
:+1:

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 4 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: I think removing "volatile" there should be okay, seems like AtomicInteger read/writes have the same memory effects as accesses on a volatile variable:

http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/atomic/package-summary.html

**Comment by jon-wei** [comment]
:+1: Cool PR, looks good to me

**Comment by jon-wei** [comment]
@navis Can you squash your commits?

**Comment by jon-wei** [comment]
@navis I'm working on some non-String typing changes to IncrementalIndex, and I want to get this PR before I submit a PR for that, can you squash the commits if you don't have any more changes?

@fjy @drcrallen @nishantmonu51 I plan on merging this after the squash, do you have any more feedback on this?

**Comment by navis** [comment]
@jon-wei Squashed

**Comment by drcrallen** [comment]
is https://github.com/navis/druid/commit/3f0c8cddc23c7154d9f323213fee9d264dc39f1f showing up for anyone else in this PR?

**Comment by drcrallen** [comment]
yeah, I think @navis you may have grabbed more commits than you meant to.

**Comment by jon-wei** [comment]
I see that CliPeon commit as well

**Comment by navis** [comment]
Sorry, my bad.

**Comment by navis** [comment]
Sorry, my bad.

**Pull request merged**